include/sundials/sundials_classview.hpp

```diff
@@ -0,0 +1,120 @@
+/* -----------------------------------------------------------------------------
+ * Programmer(s): Cody J. Balos @ LLNL
+ * -----------------------------------------------------------------------------
+ * SUNDIALS Copyright Start
+ * Copyright (c) 2025, Lawrence Livermore National Security,
+ * University of Maryland Baltimore County, and the SUNDIALS contributors.
+ * Copyright (c) 2013-2025, Lawrence Livermore National Security
+ * and Southern Methodist University.
+ * Copyright (c) 2002-2013, Lawrence Livermore National Security.
+ * All rights reserved.
+ *
+ * See the top-level LICENSE and NOTICE files for details.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ * SUNDIALS Copyright End
+ * -----------------------------------------------------------------------------
+ * Base classes for C++ implementations wrappers (views) of SUNDIALS objects.
+ * ---------------------------------------------------------------------------*/
+
+#ifndef _SUNDIALS_CLASSVIEW_HPP
+#define _SUNDIALS_CLASSVIEW_HPP
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+#include <sundials/sundials_convertibleto.hpp>
+
+namespace sundials {
+namespace experimental {
+
```

Comment:
Could consider adding something like
```c++
template <typename T, void (*D)(T)> // If C++17, use auto template arg to consolidate with other deleter below
struct ClassViewDeleter {
    void operator()(T t) const noexcept { D(t); }
};

template <typename T, void (*D)(T*)>
struct ClassViewPtrDeleter {
    void operator()(T t) const noexcept { D(&t); }
};
```
so you can simplify deleters to
```c++
using ARKodeButcherTableDeleter = ClassViewDeleter<ARKodeButcherTable, ARKodeButcherTable_Free>;
```

Reply:
This wont quite work because not all of our destroy/free routines return void. So you have to do

```cpp
template <typename T, typename Ret, Ret (*D)(T)>
struct SmartPtrDeleter {
    void operator()(T t) const noexcept { D(t); }
};

template <typename T, typename Ret, Ret (*D)(T*)>
struct SmartPtrPtrDeleter {
    void operator()(T t) const noexcept { D(&t); }
};
```

and

```
using SUNAdaptControllerDeleter =SmartPtrDeleter<SUNAdaptController, SUNErrCode, SUNAdaptController_Destroy>;
```

I don't find this particularly easier to read and it only saves 2 LOC since it adds back an include line + the using statement vs 4 lines for the struct.